### DISCUSSION TITLE
dugite-extra Node.js 14 support

### DISCUSSION THREAD

**Discussion by anavarre** [eclipse-theia/theia#10716] [General] [answered]
Hello folks!

I understand from [Theia's Developing.md prerequisites section](https://github.com/eclipse-theia/theia/blob/master/doc/Developing.md#prerequisites) that Node.js 14 is supported.

> Node.js >= 12.14.1.
For now, use Node 12.
Node 14 is supported, Node 16 should work, but you might need to run yarn --ignore-engines to get around some arbitrary constraints.

However, if I'm reading this correctly, this is not entirely accurate as there's a dependency on [dugite-extra](https://github.com/theia-ide/dugite-extra) in https://github.com/eclipse-theia/theia/blob/master/packages/git/package.json#L17 and this package is not yet compatible with 14.x as per https://github.com/theia-ide/dugite-extra/pull/54

As such, if I'm trying to build Theia on 14.x I'm getting the following error:

```
#49 187.5 error dugite-extra@0.1.14: The engine "node" is incompatible with this module. Expected version ">=10.11.0 <13". Got "14.18.3"
```

Which comes straight up from https://github.com/theia-ide/dugite-extra/blob/master/package.json#L8

Is my understanding correct? If it is, what can we do to ensure Theia will be formally compatible with Node.js 14.x since 12.x is EOL in April 2022.

Thanks!

**Comment by vince-fugnitto** [comment, marked as answer]
@anavarre thank you for the discussion, you’re right that `dugite-extra` needs to be updated in order to support newer versions, for the moment you can use `yarn —ignore-versions` (not sure if it actually causes issues to do so). 

- https://github.com/eclipse-theia/theia/blob/master/doc/Developing.md#prerequisites

I believe that `dugite-extra` did not get so much attention since it is solely used by `@theia/git` which should instead be replaced by the builtin vscode git support through the plugins (this in turn enables extensions such as gitlens from being used in the application). 

cc @thegecko I believe you might still use `@theia/git`, are you interested in the update to `dugite-extra`?

**Reply by anavarre** [reply]
Thank you, that's very helpful. Is the idea to remove `dugite-extra` entirely then? If yes, what becomes of `@theia/git`? Also, I guess I'm looking for my next step here and when you say:

> replaced by the builtin vscode git support through the plugins

I wonder what you're suggesting? Finding the best possible VSX extension? Finding the best possible VSCode extension? Any recommendations on how to proceed specifically?

**Reply by msujew** [reply]
@anavarre Git support is contributed to VSCode using a built-in extension, in particular [this one](https://open-vsx.org/extension/vscode/git/1.53.2) (using API version 1.53.2). It uses the `scm` API of VSCode (which we also support) to add git support.

A lot of users which don't use `@theia/git` rely on the built-in extension. `@theia/git` is mostly there for historical reasons, it predates the VSCode extension support by some time.

In the future, we plan to drop support for `@theia/git` entirely. Our general policy is that once a feature can be contributed by an extension, we usually deprecate it.

**Reply by anavarre** [reply]
@msujew appreciate the clarification, thanks!

**Reply by thegecko** [reply]
Thanks for the heads up @vince-fugnitto 

We do use `@theia/git` but are still a few versions behind.

cc @westbury

**Comment by anavarre** [comment]
@msujew many thanks for merging https://github.com/theia-ide/dugite-extra/pull/54 - Can we expect a new tag to be created so in return Theia can ship with an updated version and officially claim Node.js 14 compatibility?

**Reply by msujew** [reply]
I don't think we have plans to release a minor update for this (creating a release is considerable overhead for our team). Luckily, you don't need a new version of Theia to enjoy Node 14 "compatibility". You can simply add a resolutions section to the __root__ `package.json` of your Theia project:

```json
"resolutions": {
  "find-git-exec": "0.0.4",
  "dugite-extra": "0.1.15"
}
```

You can remove this again after updating to Februaries' Theia version.

**Reply by anavarre** [reply]
I failed to clarify I was talking about https://www.npmjs.com/package/dugite-extra and I just noticed you created a new tag 3 days ago which resolves my concern fully! Thanks again.

**Reply by msujew** [reply]
You're welcome :) Using the new tag in Theia was already merged a few days ago as well, see https://github.com/eclipse-theia/theia/pull/10722.
